### PR TITLE
Update image generation process with more consistent naming

### PR DESCRIPTION
**Pull request by youngnick**
Signed-off-by: Nick Young <ynick@vmware.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR adds some tooling around building admission-server images, so that we end up with:
- all images built and pushed are labelled with the standard Prow version string, like `v20220209-v0.4.0-114-g36b4551`.
- Images built for a semver branch or tag are also labelled with the branch or tag name
- Images built for a semver branch or tag that don't have a prerelease marker (like `-rc2`, `-beta1` or similar are labelled `latest`.

This will mean that we can now generate consistent manifest YAML for the admission-server when we release a bundle by tagging a branch with a semver tag, since the images will be available by `v0.4.1` or whatever.

I'll leave changing out the build to a multiarch one (#627) for another PR. Having it in a script should make that easier though.

**Which issue(s) this PR fixes**:
Updates #916 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by youngnick** [comment]
I _think_ I've got the interactions between Prow, cloud build, and our Makefile correct, but looking forward to reviews around that. I'm not sure how to test there.

File: hack/build-and-push.sh [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+if [[ $VERSION_TAG != $GIT_TAG ]]
+then
+	echo docker tag ${REGISTRY}/admission-server:${GIT_TAG} ${REGISTRY}/admission-server:${VERSION_TAG}
+	echo docker push ${REGISTRY}/admission-server:${VERSION_TAG}
+fi
+
+if [[ $LATEST == true ]]
+then
+	echo docker tag ${REGISTRY}/admission-server:${GIT_TAG} ${REGISTRY}/admission-server:latest
+	echo docker push ${REGISTRY}/admission-server:latest
```
Comment (robscott) [inline hack/build-and-push.sh:89]: Nit: Hard to tell in GitHub, but this indentation looks inconsistent with what's further above in the file.

File: Makefile [hunk 2]
```diff
@@ -12,11 +12,19 @@
...
+
 # Enable Go modules.
-export GO111MODULE=on
+GO111MODULE=on
 
 REGISTRY ?= gcr.io/k8s-staging-gateway-api
-TAG ?= dev
+
+# These are overridden by cloudbuild.yaml when run by Prow.
+GIT_TAG ?= dev
```
Comment (robscott) [inline Makefile:25]: Do you have a reference to what these are overridden by?

File: Makefile [hunk 3]
```diff
@@ -12,11 +12,19 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# We need all the Make variables exported as env vars.
+# Note that the ?= operator works regardless.
+.EXPORT_ALL_VARIABLES:
```
Comment (robscott) [inline Makefile:17]: It would be helpful to note that this is specifically for `release-staging`. I'm not familiar with best practices here, but would it be preferable to explicitly export each individual variable instead so we don't end up with accidental exports in the future?

**Review by robscott — COMMENTED** [review]
Thanks for the work on this @youngnick! Most LGTM, also not sure how to test this.

File: hack/build-and-push.sh (hunk 1 above)
Comment (youngnick) [reply hack/build-and-push.sh:89]: Thanks, those were tabs, not spaces. Fixed.

File: Makefile (hunk 2 above)
Comment (youngnick) [reply Makefile:25]: I've added some reference into the Makefile here.

File: Makefile (hunk 3 above)
Comment (youngnick) [reply Makefile:17]: Good point, I've removed and explicitly exported the required env vars.

**Comment by youngnick** [comment]
I spoke to a couple of other VMware folks, and they were both of the "suck it and see" school when it comes to testing these changes.

I made sure that this change doesn't need any test-infra changes, so, assuming that `cloudbuild.yaml` works as advertised, it all should just work (tm).

**Comment by youngnick** [comment]
Okay, some testing of the script itself.

tl;dr The script appears to work.

First, I changed all the actual docker commands to `echo` the command instead:

```bash
echo docker build --build-arg COMMIT=${BINARY_VERSION} --build-arg TAG=${VERSION_TAG} \
  			-t ${REGISTRY}/admission-server:${GIT_TAG} .

echo docker push ${REGISTRY}/admission-server:${GIT_TAG}

```

Now, let's check the various combinations of variables. Each of these should set the binary version in the docker build command to `$COMMIT`.

This combination should push only the `$GIT_TAG` label.
```bash
export REGISTRY=gcr.io/k8s-staging-gateway-api
export GIT_TAG=v0.4.0-rc1-210-gcd7d431
export BASE_REF=master
export COMMIT=$(git rev-parse --short HEAD)
./hack/build-and-push.sh
```
gives
```bash
docker build --build-arg COMMIT=cd7d431 --build-arg TAG=v0.4.0-rc1-210-gcd7d431 -t gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431 .
docker push gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431
```

If `$BASE_REF` is semver with a prerelase marker, we should get the semver set as the change, but no `:latest` label on the image.
```bash
export REGISTRY=gcr.io/k8s-staging-gateway-api
export GIT_TAG=v0.4.0-rc1-210-gcd7d431
export BASE_REF=v0.5.0-rc1
export COMMIT=$(git rev-parse --short HEAD)
./hack/build-and-push.sh
```
gives
```bash
docker build --build-arg COMMIT=v0.5.0-rc1 --build-arg TAG=v0.5.0-rc1 -t gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431 .
docker push gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431
docker tag gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431 gcr.io/k8s-staging-gateway-api/admission-server:v0.5.0-rc1
docker push gcr.io/k8s-staging-gateway-api/admission-server:v0.5.0-rc1
```

If `$BASE_REF` is semver with no prerelase marker, we should get the semver set as the change, with a `:latest` label on the image as well.
```bash
export REGISTRY=gcr.io/k8s-staging-gateway-api
export GIT_TAG=v0.4.0-rc1-210-gcd7d431
export BASE_REF=v0.5.0
export COMMIT=$(git rev-parse --short HEAD)
./hack/build-and-push.sh
```
gives
```bash
docker build --build-arg COMMIT=v0.5.0 --build-arg TAG=v0.5.0 -t gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431 .
docker push gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431
docker tag gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431 gcr.io/k8s-staging-gateway-api/admission-server:v0.5.0
docker push gcr.io/k8s-staging-gateway-api/admission-server:v0.5.0
docker tag gcr.io/k8s-staging-gateway-api/admission-server:v0.4.0-rc1-210-gcd7d431 gcr.io/k8s-staging-gateway-api/admission-server:latest
docker push gcr.io/k8s-staging-gateway-api/admission-server:latest
```



Now, to test actually building and pushing:
```bash
export REGISTRY=docker.io/youngnick
export GIT_TAG=v0.4.0-rc1-210-gcd7d431
export BASE_REF=v0.5.0
export COMMIT=$(git rev-parse --short HEAD)
./hack/build-and-push.sh
```
gives
```bash
Sending build context to Docker daemon  92.86MB
Step 1/12 : FROM golang:1.16 AS build-env
 ---> 1390628e9890
Step 2/12 : RUN mkdir -p /go/src/sig.k8s.io/gateway-api
 ---> Using cache
 ---> 3a9ddf207c7c
Step 3/12 : WORKDIR /go/src/sig.k8s.io/gateway-api
 ---> Using cache
 ---> 0545b77e58ba
Step 4/12 : COPY  . .
 ---> Using cache
 ---> 642635845bfc
Step 5/12 : ARG TAG
 ---> Using cache
 ---> ba71ac6dabb7
Step 6/12 : ARG COMMIT
 ---> Using cache
 ---> 2a733e843eb2
Step 7/12 : RUN CGO_ENABLED=0 GOARCH=amd64 GOOS=linux go build -a -o gateway-api-webhook       -ldflags "-s -w -X main.VERSION=$TAG -X main.COMMIT=$COMMIT" ./cmd/admission
 ---> Using cache
 ---> 69c7836b37cf
Step 8/12 : FROM gcr.io/distroless/static:nonroot
 ---> 5bc88c360b3e
Step 9/12 : WORKDIR /
 ---> Using cache
 ---> d4f8dcf1018d
Step 10/12 : COPY --from=build-env /go/src/sig.k8s.io/gateway-api/gateway-api-webhook .
 ---> Using cache
 ---> 062d45582ffa
Step 11/12 : USER 65532
 ---> Using cache
 ---> 95ef5fcfce26
Step 12/12 : ENTRYPOINT ["/gateway-api-webhook"]
 ---> Using cache
 ---> 189d27b922e4
Successfully built 189d27b922e4
Successfully tagged youngnick/admission-server:v0.4.0-rc1-210-gcd7d431
The push refers to repository [docker.io/youngnick/admission-server]
34bca3bf2868: Pushed 
5b1fa8e3e100: Pushed 
v0.4.0-rc1-210-gcd7d431: digest: sha256:c3e2a245095f1e97c097b7345101eafda521f1819cae75a62700c192c04ffabd size: 738
The push refers to repository [docker.io/youngnick/admission-server]
34bca3bf2868: Layer already exists 
5b1fa8e3e100: Layer already exists 
v0.5.0: digest: sha256:c3e2a245095f1e97c097b7345101eafda521f1819cae75a62700c192c04ffabd size: 738
The push refers to repository [docker.io/youngnick/admission-server]
34bca3bf2868: Layer already exists 
5b1fa8e3e100: Layer already exists 
latest: digest: sha256:c3e2a245095f1e97c097b7345101eafda521f1819cae75a62700c192c04ffabd size: 738
```

**Comment by youngnick** [comment]
I had a crack at replicating what Prow does with `gcloud builds submit`, but Prow does some extra magic with pulling the source for you, and I have no idea where to start there.

overall tl;dr - I'm confident this will do what it intends.

File: Makefile [hunk 4]
```diff
@@ -12,12 +12,33 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# We need all the Make variables exported as env vars.
+# Note that the ?= operator works regardless.
+
 # Enable Go modules.
-export GO111MODULE=on
+GO111MODULE=on
```
Comment (robscott) [inline Makefile:19]: Was this change intentional?

**Review by robscott — COMMENTED** [review]
One question but otherwise this LGTM, thanks @youngnick!

File: Makefile (hunk 4 above)
Comment (youngnick) [reply Makefile:19]: Nope, reverted, thanks!

**Comment by robscott** [comment]
Thanks!

/lgtm
